spec/controllers/users_controller_spec.rb

```diff
@@ -0,0 +1,98 @@
+require 'rails_helper'
+
+def generate_unique_email(base_email)
+  random_string = SecureRandom.hex(4) # Generate a random string of length 8
+  domain = base_email.split('@').last
+  modified_email = "#{base_email.gsub('@' + domain, '')}+#{random_string}@#{domain}"
+  modified_email.downcase
+end
+
+RSpec.describe UsersController, type: :controller do
+  include SessionsHelper
+
+  let(:valid_attributes) do
+    {
+      name: 'John Doe',
+      email: generate_unique_email('johndoe@yahoo.com'),
+      password: 'password123',
+      password_confirmation: 'password123'
+    }
+  end
+
+  describe 'POST #create' do
```

Comment:
There are multiple places where you do the same action:
```ruby
post :create, params: { user: valid_attributes }
```

So what you could do here, is extract that into a `subject`. Your code will look like:
```ruby
describe 'POST #create' do
  let(:valid_create_attributes) do
    {
      name: 'John Doe',
      email: generate_unique_email('johndoe@yahoo.com'),
      password: 'password123',
      password_confirmation: 'password123'
    }
  end

  subject { post :create, params: { user: valid_attributes } }

  context 'with valid attributes' do
    it 'creates a new user' do
      expect { subject }.to change(User, :count).by(1)
    end

    it 'sends email activation to the user' do
      expect { subject }.to change(ActionMailer::Base.deliveries, :count).by(1)
    end

    # all tests where the attributes are valid
  end

  # in the context where attributes are not valid
  context 'with valid attributes' do
    # here you modify, add or remove any attribute you want, to make the final params structure to be invalid
    before { valid_create_attributes.merge!({}) }

    # add all the tests with this format of attributes, where they are invalid
  end
end
```

The `subject` is basically an action, with given params. In certain contexts, you will have a `before` action, that will modify the `valid_attributes` variable. So the subject will end up doing the same action, but with the modified `valid_attributes`. So the action (calling `post :create`) will be the same, but the params can be different